usr/lib/cca_stdll/cca_specific.c

```diff
@@ -239,6 +240,300 @@ static const CK_ULONG cca_mech_list_len =
+    case CKA_DERIVE: typestr = "CKA_DERIVE"; break;
+    case CKA_START_DATE: typestr = "CKA_START_DATE"; break;
+    case CKA_END_DATE: typestr = "CKA_END_DATE"; break;
+    case CKA_MODULUS: typestr = "CKA_MODULUS"; break;
+    case CKA_MODULUS_BITS: typestr = "CKA_MODULUS_BITS"; break;
+    case CKA_PUBLIC_EXPONENT: typestr = "CKA_PUBLIC_EXPONENT"; break;
+    case CKA_PRIVATE_EXPONENT: typestr = "CKA_PRIVATE_EXPONENT"; break;
+    case CKA_PRIME_1: typestr = "CKA_PRIME_1"; break;
+    case CKA_PRIME_2: typestr = "CKA_PRIME_2"; break;
+    case CKA_EXPONENT_1: typestr = "CKA_EXPONENT_1"; break;
+    case CKA_EXPONENT_2: typestr = "CKA_EXPONENT_2"; break;
+    case CKA_COEFFICIENT: typestr = "CKA_COEFFICIENT"; break;
+    case CKA_PRIME: typestr = "CKA_PRIME"; break;
+    case CKA_SUBPRIME: typestr = "CKA_SUBPRIME"; break;
+    case CKA_BASE: typestr = "CKA_BASE"; break;
+    case CKA_PRIME_BITS: typestr = "CKA_PRIME_BITS"; break;
+    case CKA_SUBPRIME_BITS: typestr = "CKA_SUBPRIME_BITS"; break;
+    case CKA_VALUE_BITS: typestr = "CKA_VALUE_BITS"; break;
+    case CKA_VALUE_LEN: typestr = "CKA_VALUE_LEN"; break;
+    case CKA_EXTRACTABLE: typestr = "CKA_EXTRACTABLE"; break;
+    case CKA_LOCAL: typestr = "CKA_LOCAL"; break;
+    case CKA_NEVER_EXTRACTABLE: typestr = "CKA_NEVER_EXTRACTABLE"; break;
+    case CKA_ALWAYS_SENSITIVE: typestr = "CKA_ALWAYS_SENSITIVE"; break;
+    case CKA_KEY_GEN_MECHANISM: typestr = "CKA_KEY_GEN_MECHANISM"; break;
+    case CKA_MODIFIABLE: typestr = "CKA_MODIFIABLE"; break;
+    case CKA_EC_PARAMS: typestr = "CKA_EC_PARAMS"; break;
+    case CKA_EC_POINT: typestr = "CKA_EC_POINT"; break;
+    case CKA_ALWAYS_AUTHENTICATE: typestr = "CKA_ALWAYS_AUTHENTICATE"; break;
+    case CKA_IBM_OPAQUE: typestr = "CKA_IBM_OPAQUE"; break;
+    default:
+        sprintf(tbuf, "unknown type 0x%08lx", a->type);
+        typestr = tbuf;
+        break;
+    }
+
+    switch (a->ulValueLen) {
+    case 0:
+        TRACE_DEBUG("  %s: len=0 pValue=%p\n", typestr, a->pValue);
+        break;
+    case 1:
+        TRACE_DEBUG("  %s: len=%lu value=0x%02hhx\n",
+                    typestr, a->ulValueLen, *((uint8_t *)(a->pValue)));
+        break;
+    case 2:
+        TRACE_DEBUG("  %s: len=%lu value=0x%04hx\n",
+                    typestr, a->ulValueLen, *((uint16_t *)(a->pValue)));
+        break;
+    case 4:
+        TRACE_DEBUG("  %s: len=%lu value=0x%08x\n",
+                    typestr, a->ulValueLen, *((uint32_t *)(a->pValue)));
+        break;
+    case 8:
+        TRACE_DEBUG("  %s: len=%lu value=0x%016lx\n",
+                    typestr, a->ulValueLen, *((uint64_t *)(a->pValue)));
+        break;
+    default:
+        TRACE_DEBUG("  %s: len=%lu value:\n", typestr, a->ulValueLen);
+        dump_mem("  ", a->pValue, a->ulValueLen);
+        break;
+    }
+}
+
+/* Debug function: dump list of attribues from a template via TRACE_DEBUG */
+static void dump_template(TEMPLATE *tmpl)
```

Comment:
move to template.c, or p11util.c ?

Reply:
done, will be part of the next PR update